subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java

```diff
@@ -44,6 +47,21 @@
      */
     void setConsoleOutput(ConsoleOutput consoleOutput);
 
+    /**
+     * Specifies if deprecation warnings should be written to the console.
+     * Defaults to {@link WarningType#Summary}
+     * @since 4.5
+     */
+    @Incubating
+    Set<WarningType> getWarningTypes();
```

Comment:
I somewhat don't understand how `NONE` fits into a set of warnings. Would that mean that, if I use the invocation
```
./gradlew --warnings all --warnings none
```
then this set is `[NONE, ALL]` and all warnings would be displayed?
I also wouldn't be able to override a setting of `org.gradle.warnings=all` in my `gradle.properties` by passing `--warnings none` on the command line, right?

Or am I understanding something wrong here?

Reply:
Well, in my understanding, this should be "undefined behavior" currently. This usage is confusing for user, just like `./gradlew --info --debug --error`. Currently we only have `--all` and `--none`, they shouldn't show up at the same time because it's mutually exclusive. In the future, it could be something like `--warnings deprecations --warnings compiler`.

Here we use `Set` because this is a public API, we should not change it after publication.

Reply:
I think the use case of setting `org.gradle.warnings=all` in `gradle.properties` and then running with `--warnings none` is a legitimate use case.

I think that `deprecation` and `compiler` are something completely different to `all`, `none` and `summary`. The latter three determine what to which relevant warnings, i.e. show them individually, don't show them, or show their number.
`deprecation` and `compiler` on the other hand are which types of warnings to take into account.

So I guess something like `--show-warnings all,summary` makes sense as would `--show-warnings summary --warnings compiler,deprecations`.

I guess `none` would evaluate to an empty set for how to show the warnings instead of a set with the element `NONE`. `all` would then probably be the set with `INDIVIDUAL,SUMMARY` (if that is what we want).

WDYT?

Reply:
I think we could do what the java compiler does. It's one option with 2 special values ("all" and "none"). The special values overwrite anything that was there before. The other values are simply additive.

See https://docs.oracle.com/javase/7/docs/technotes/tools/windows/javac.html#xlintwarnings

Reply:
@oehme That sounds good for `all` or `none` but it doesn't fit with `summary` I guess.
The Java compiler also uses categories which can be activated by `-Xlint:all`, `-Xlint:none` or individually by `-Xlint:<category>`, but not different kinds of reporting (e.g. our `summary`).

It uses `-Xmaxwarns number` to select the number of warnings to show. That is similar to our `summary` option I guess.

Reply:
I see what you mean, that split makes sense.

Reply:
Thanks @wolfs . Personally I don't like `--show-warnings` and `--warnings` because their names are so similar that users might be puzzled. I think we could have `--warnings` for `all`/`none`/`<category>` for now:

- By default, this set is empty, all categories of warnings are suppressed and a summary will be rendered.
- If special value `none`/`all` is specified, all other values are overwritten, just like @oehme said. This means it's allowed to have `org.gradle.warnings=all` and cmd args `--warnings=none`
- `--warnings=all --warnings=none` is undefined.

Reply:
I am not very fixed on what the options are called. I would very like to see both concepts modelled differently, though.
IIUC, we currently only implement the way how to show the warnings.
So it is OK to only have the switch for the how to present the warnings, with possible values `summary`, `full` or `none`. That would not be a `Set` but a single value.
Then we can later add support for categories, like `deprecation`, etc.
WDYT?

Reply:
Before reading this thread I also wondering why we can provide a Set of warnings. Now this makes more sense. I agree with @wolfs that we should separate warning type from the warning category/level.

Reply:
Thanks @wolfs @bmuschko @oehme I've already done this. Please take a look at it again when you're free.